src/metabolic_ninja/worker/main.py

```diff
@@ -0,0 +1,154 @@
+# that anyway).
+sentry_sdk.integrations.logging.ignore_logger(
+    "pika.adapters.utils.io_services_utils"
+)
+sentry_sdk.integrations.logging.ignore_logger(
+    "pika.adapters.utils.connection_workflow"
+)
+sentry_sdk.integrations.logging.ignore_logger(
+    "pika.adapters.blocking_connection"
+)
+
+# Whenever a job is received, the work on it will be started in a new thread. This is
+# to allow the RabbitMQ i/o loop to do its thing (like sending hearbeats to the server
+# to keep the connection alive). Started threads are stored in this list, so that we can
+# wait for the threads to complete when terminating the application.
+worker_threads = []
+
+
+def on_message(channel, method_frame, header_frame, body, connection):
+    """Handle new messages from RabbitMQ."""
+    logger.debug(f"Received message {method_frame.delivery_tag}")
+    # Start the work in a separate thread, to avoid blocking the pika i/o loop.
+    thread = threading.Thread(
+        target=tasks.design,
+        args=(
+            connection,
+            channel,
+            method_frame.delivery_tag,
+            body,
+            ack_message,
+        ),
+    )
+    thread.start()
+    worker_threads.append(thread)
+
+
+def ack_message(channel, delivery_tag):
+    """Acknowledge a finished job."""
+    # If the channel was closed for some reason, ignore.
+    logger.debug(f"ACKing message {delivery_tag}")
+    if channel.is_open:
+        channel.basic_ack(delivery_tag)
+
+
+def on_terminate(channel, signum, frame):
+    """SIGTERM signal handler to terminate the application."""
+    logger.debug(f"Caught SIGTERM, cancelling consumption")
+    channel.stop_consuming()
+
+
+def main():
+    logger.debug("Establishing connection and declaring task queue")
+    try:
+        connection = pika.BlockingConnection(
+            pika.ConnectionParameters(host=os.environ["RABBITMQ_HOST"])
+        )
+    except pika.exceptions.AMQPConnectionError:
+        # Connection failed - exit and rely on the environment to restart the
+        # application. Don't let the exception bubble though; we don't need a Sentry
+        # notification for this.
+        logger.warning(
+            "AMQPConnectionError: Cannot connect to RabbitMQ. Aborting."
+        )
+        sys.exit(-1)
+
+    channel = connection.channel()
+    channel.queue_declare(queue="jobs", durable=True)
+    # Prefetch only a single message, to ensure messages aren't sent to busy
+    # workers.
+    channel.basic_qos(prefetch_count=1)
+    # Pass the connection to the message callback - it'll be needed later to ACK
+    # messages.
+    callback = functools.partial(on_message, connection=connection)
+    channel.basic_consume(queue="jobs", on_message_callback=callback)
+
+    # Register the signal handler
+    signal.signal(signal.SIGTERM, functools.partial(on_terminate, channel))
+
+    logger.info("Ready for action, waiting for messages from RabbitMQ")
+    channel.start_consuming()
```

Comment:
I'm assuming that this starts some sort of loop that polls the queue. (Don't know the pika library.) I'm wondering if threads should be joined inside of that loop somehow rather than cleaning up only at the very end? A worker might otherwise hit an open thread limit?

Reply:
> I'm assuming that this starts some sort of loop that polls the queue.

Indeed, that's the i/o loop which communicates with the rabbitmq channel.

> I'm wondering if threads should be joined inside of that loop somehow rather than cleaning up only at the very end? A worker might otherwise hit an open thread limit?

There will only be one thread running at any time because we're only prefetching a single message, and acknowledging it at the very end of the thread. ([Read more about prefetching here](https://www.rabbitmq.com/confirms.html#channel-qos-prefetch).)

The only reason we're spawning these threads is to not block the i/o loop and allow it to send ping/heartbeats on the channel concurrently while workers are working.

We still need to join after the consumption exist, in case the application is terminated. The `on_terminate` handler will stop consumption but worker threads may still be running and expect the channel to be open. (We don't have explicit handling of `TERM`/`KILL` signals sent to the child process, so we're currently relying on cameo's behaviour).

Reply:
Cheers, thank you for the explanation.